pkg/zng/docs/spec.md

```diff
@@ -1,92 +1,366 @@
+type 24 = record[_path:string,ts:time,uid:string,id:record[orig_h:ip,orig_p:port,resp_h:addr,resp_p:port]...
+value[23]: conn;1425565514.419939;CogZFI3py5JsFZGik;[192.168.1.1:;80/tcp;192.168.1.2;8080;]...
+value[24]: dns;1425565514.419987;CogZFI3py5JsFZGik;[192.168.1.1:;5353/udp;192.168.1.2;5353;]...
+```
+Note that the value encoding need not refer to the field names and types as that is
+completely captured by the type code.  Values merely encode the value
+information consistent with the referenced type code.
+
+## ZNG Binary Format (BZNG)
+
+In BZNG, the data model defined above is serialized into a stream of bytes.
+The byte stream encodes a sequence of messages.
 
-ZSON is a UTF-8 encoded stream of "lines" where each line is terminated by a
-newline.  Each line is either a directive or a value.
+Each message is prefixed with a single-byte header code.  The upper bit of
+the header code indicates whether the message is a control message (1)
+or a value (0).
 
-Directives and values, in turn, come in two flavors: regular and legacy.
-Thus, there are four types of lines:
-* regular directives,
-* regular values,
-* legacy directives, and
-* legacy values.
+In the case of control message, the lower 7 bits define the control type.
+Control type 0 is a "type definition" or typedef described below.
 
-Any line that begins with character `#` is a directive.
-All other lines are values.  If a value line begins with `#`, then this
-character must be escaped.
-(Such lines can only exist as legacy values and not regular values
-as regular values always begin with an integer descriptor.)
+Control code 1 is reserved for the ordering hint.  All other control codes
+are available to higher-layer protocols for carrying protocol-specific payloads
+embedded in the ZNG stream.
 
-Directives indicate how subsequent values in the ZSON stream are interpreted.
-A value is a regular value if the most recent preceding directive in the stream
-is a regular directive; otherwise, it is a legacy value.
+Control messages may be used informatively and shall be
+ignored by any data receivers.  The message can be any UTF-8 string.
+Control payloads are guaranteed to be preserved
+in order within the stream and presented to higher layer components through
+any ZNG streaming API.  In this way, senders and receivers of ZNG can embed
+protocol directives as ZNG control payloads rather than defining additional
+encapsulating protocols.  See the
+[zng-over-http](zng-over-http.md) protocol for an example.
+
+### BZNG Types
+
+Following a header byte of 0x80 is a "type definition" (typedef) that binds
+"the next available" integer type code to a type encoding.  Type codes
+begin at the value 23 and increase by one for each typedef. This bindings
+are scoped to the stream in which the typedef occurs.
+
+Type codes for the "scalar types" are predefined as follows:
+
+| Type     | Code |
+|----------|------|
+| bool     |   0  |
```

Comment:
I'm partial to making liberal use of fixed width stuff like `bool` and `0` when making reference to formal identifiers and code as they're used within a format/protocol.

Reply:
I don't understand.

Reply:
I was referring to making the entries fixed width, e.g.:

| Type     | Code |
|----------|------|
|`bool`    |  `0` |
...

Reply:
got it